From: Robin Holt <holt@sgi.com>
Subject: EPOLL: fix max_user_watches overflow
References: bnc#652391
Patch-mainline: never

On a 16TB machine, max_user_watches has an integer overflow.  Convert it
to use a long and handle the associated fallout.

Fixed by 52bd19f7691b2ea6433aef0ef94c08c57efd7e79 upstream.

Signed-off-by: Robin Holt <holt@sgi.com>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>

---
 fs/eventpoll.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/fs/eventpoll.c
+++ b/fs/eventpoll.c
@@ -1510,8 +1510,8 @@ static int __init eventpoll_init(void)
 	/*
 	 * Allows top 4% of lomem to be allocated for epoll watches (per user).
 	 */
-	max_user_watches = (((si.totalram - si.totalhigh) / 25) << PAGE_SHIFT) /
-		EP_ITEM_COST;
+	max_user_watches = min((unsigned long)INT_MAX, (((si.totalram - si.totalhigh) / 25) << PAGE_SHIFT) /
+		EP_ITEM_COST);
 
 	/*
 	 * Initialize the structure used to perform epoll file descriptor
